src/Data/Text/Utf8/Unlower.hs

```diff
@@ -26,13 +26,19 @@ import qualified Data.List as List
 --
 unlowerCodePoint :: Char -> [Char]
 unlowerCodePoint =
-  \c -> maybe [] id $ HashMap.lookup c completeMapping
+  \c -> maybe [c] id $ IntMap.lookup (Char.ord c) unlowerings
+
+-- | This map contains all the unlowerings for which the result is not just a singleton with the
+-- input character. It's marked NOINLINE to make sure that it only gets constructed once.
+unlowerings :: IntMap.IntMap [Char]
```

Comment:
Does using an `IntMap` actually provide a performance benefit? My earlier comment was mainly meant as: this *could* be faster since it doesn't box the key, but in practice it might have other downsides.

Reply:
I didn't measure it but thought it would surely be faster. Turns out that hashmap is actually faster :grimacing:. At least for the ascii range (where you get relatively many lookup hits I suppose).

```
unlower intmap A..Z      513.9 ns   (508.1 ns .. 521.0 ns)
unlower intmap a..z      576.9 ns   (568.3 ns .. 587.3 ns)
unlower intmap ascii     1.839 μs   (1.822 μs .. 1.864 μs)
unlower intmap min..max  7.119 ms   (7.076 ms .. 7.168 ms)
unlower hashmap A..Z     330.1 ns   (328.0 ns .. 332.2 ns)
unlower hashmap a..z     423.1 ns   (420.9 ns .. 425.8 ns)
unlower hashmap ascii    1.094 μs   (1.087 μs .. 1.101 μs)
unlower hashmap min..max 9.427 ms   (9.318 ms .. 9.530 ms)
```

Reply:
*Meten is weten*

One other thing that comes to mind (maybe for a follow-up PR then) is employing a separate lookup-table for the characters < 256, like in the search automata.